keel-core/src/test/kotlin/com/netflix/spinnaker/keel/actuation/IntermittentFailureTests.kt

```diff
@@ -0,0 +1,170 @@
+import java.time.Clock
+import kotlinx.coroutines.runBlocking
+import org.springframework.context.ApplicationEventPublisher
+
+/**
+ * This class tests a specific scenario:
+ * (1) a resource has a diff which we cannot fix, so it is vetoed.
+ * (2) that diff goes away (it was an intermittent failure of another system, maybe) and we say it's happy
+ * (3) that same diff comes back, and we still try again because it went from happy to diff.
+ *
+ * We see this happen sometimes with resources because all systems fail sometimes. This test makes sure
+ * we can recover from those failures.
+ */
+class IntermittentFailureTests : JUnit5Minutests {
+
+  class Fixture {
+    val resourceRepository = InMemoryResourceRepository()
+    val artifactRepository = mockk<ArtifactRepository>()
+    val deliveryConfigRepository = mockk<DeliveryConfigRepository>()
+    val diffFingerprintRepository = InMemoryDiffFingerprintRepository()
+    val actuationPauser: ActuationPauser = mockk() {
+      coEvery { isPaused(any<String>()) } returns false
+      coEvery { isPaused(any<Resource<*>>()) } returns false
+    }
+    val plugin1 = mockk<ResourceHandler<DummyResourceSpec, DummyResourceSpec>>(relaxUnitFun = true)
+    val plugin2 = mockk<ResourceHandler<DummyResourceSpec, DummyResourceSpec>>(relaxUnitFun = true)
+    val publisher = mockk<ApplicationEventPublisher>(relaxUnitFun = true)
+    val clock = MutableClock()
+    val vetoRepository = InMemoryUnhappyVetoRepository(clock)
+
+    val dynamicConfigService: DynamicConfigService = mockk(relaxUnitFun = true) {
+      every {
+        // mimicking how a cluster is set up
+        getConfig(Int::class.java, "veto.unhappy.max-diff-count", any())
+      } returns 2
+      every {
+        // mimicking how a cluster is set up
+        getConfig(String::class.java, "veto.unhappy.waiting-time", any())
+      } returns "PT0S" // Duration.ZERO
+    }
+
+    val veto = UnhappyVeto(
+      resourceRepository,
+      diffFingerprintRepository,
+      vetoRepository,
+      dynamicConfigService,
+      "PT10M"
+    )
+    val vetoEnforcer = VetoEnforcer(listOf(veto))
+    val subject = ResourceActuator(
+      resourceRepository,
+      artifactRepository,
+      deliveryConfigRepository,
+      diffFingerprintRepository,
+      listOf(plugin1, plugin2),
+      actuationPauser,
+      vetoEnforcer,
+      publisher,
+      Clock.systemUTC())
+    val desired = DummyResourceSpec(data = "fnord")
+    val current = DummyResourceSpec()
+    val diff = DefaultResourceDiff(desired, current)
+  }
+
+  fun tests() = rootContext<Fixture> {
+
+    fixture { Fixture() }
+
+    before {
+      every { plugin1.name } returns "plugin1"
+      every { plugin1.supportedKind } returns SupportedKind(parseKind("plugin1/foo@v1"), DummyResourceSpec::class.java)
+      every { plugin2.name } returns "plugin2"
+      every { plugin2.supportedKind } returns SupportedKind(parseKind("plugin2/bar@v1"), DummyResourceSpec::class.java)
```

Comment:
Since these apply to all tests, any reason not to have the `every` statements above where you declare the mocks, like you did for other mocks?

Reply:
yep changed